server/src/controllers/grades.ts

```diff
@@ -0,0 +1,158 @@
+// SPDX-FileCopyrightText: 2023 The Aalto Grades Developers
+//
+// SPDX-License-Identifier: MIT
+
+import { Request, Response } from 'express';
+
+import Attainable from '../database/models/attainable';
+import UserAttainmentGrade from '../database/models/userAttainmentGrade';
+
+import { ApiError } from '../types/error';
+import {
+  CalculationResult,
+  Formula,
+  formulaChecker,
+  FormulaNode,
+  FormulaParams,
+  getFormula,
+  Status
+} from '../types/formulas';
+import { idSchema } from '../types/general';
+import { HttpCode } from '../types/httpCode';
+
+async function calculate(
+  tree: FormulaNode,
+  presetPoints: Map<FormulaNode, number>
+): Promise<CalculationResult> {
+  if (presetPoints.has(tree)) {
```

Comment:
Comment to explain this is to handle cases where a teacher has manually specified a grade for an attainment.